### PR TITLE
perf: fp16 normalization speedups for PLKSR, NAFNet, SPAN

### PR DESCRIPTION
**Pull request by NevermindNilas**
## Summary

Three independent fp16 inference fixes. The common thread: PyTorch's fp16
normalization kernels are either slow or numerically unsound for these shapes,
so the fix is to keep the **normalization statistics in fp32** while leaving the
heavy elementwise work in the input dtype.

All changes are numerically equivalent (within fp16 noise), remain
`torch.onnx`-exportable, and are TensorRT-neutral (no engine regression).
Benchmarks below were measured on an RTX 3090; full table and a reproducible
script are in `scripts/bench_fp16_norm.{py,md}`.

### PLKSR (`RealPLKSR`)

`nn.GroupNorm` in fp16 hits a slow kernel (~6-8x slower than a manual reduction
for these shapes) and forces a contiguous copy every block. Replaced with a
`FastGroupNorm` drop-in (same parameters / `state_dict` keys) that computes
fp32-accumulated centered statistics, normalizes in fp32, then applies the
affine in the input dtype. **+31-36% fps**, PSNR/SSIM identical, reserved VRAM unchanged.

### NAFNet (`LayerNorm2d`)

The custom-autograd `LayerNormFunction` overflowed in fp16 -- `(x - mu)**2`
exceeds 65504, and `eps=1e-6` falls below fp16 resolution -> `sqrt(~0)` -> garbage
output (PSNR ~4 dB), which forces fp32. Replaced with plain fp32-stat ops, which
makes fp16 correct (PSNR ~76-80 dB) and unlocks fp16. **+46-47% fps vs the previously-required fp32.**

### SPAN (`Conv3XC`)

`forward` called `update_params()` (re-param weight fusion) on every forward in
eval mode, across ~20 instances. It only reads unchanging training weights, so
it is idempotent -- fused once now (lazily, after load) instead of every frame.
Output is bit-identical. **+33% @360p, +11% @720p, +7% @1080p.**

### PLKSR -- `real-plksr`
| res | dtype | fps before | fps after | Delta | PSNR | SSIM | VMAF |
|-----|-------|-----------:|----------:|------:|-----:|------|-----:|
| 360p  | fp16 | 5.99 | **7.83** | +31% | 79.6 | 0.99999 | -- |
| 360p  | fp32 | 3.63 | **4.31** | +19% | 79.6 | 0.99999 | -- |
| 720p  | fp16 | 1.47 | **1.92** | +31% | 79.3 | 0.99999 | 98.1 |
| 720p  | fp32 | 0.89 | **1.06** | +19% | 79.3 | 0.99999 | -- |
| 1080p | fp16 | 0.63 | **0.84** | +33% | 79.4 | 0.99999 | -- |

### NAFNet -- `NAFNet-GoPro-width64` (fp16 was broken before; fp32 was the usable baseline)
| res | mode | fps | PSNR | SSIM | VMAF |
|-----|------|----:|-----:|------|-----:|
| 360p  | fp32 (before) | 15.4 | -- | -- | -- |
| 360p  | fp16 (after)  | **22.6** (+47%) | 75.9 | 0.99999 | -- |
| 720p  | fp32 (before) | 4.08 | -- | -- | -- |
| 720p  | fp16 (after)  | **5.97** (+46%) | 76.6 | 0.99999 | 98.2 |
| 1080p | fp16 (after)  | **2.65** | 80.6 | 0.99999 | -- |

Original fp16 (reference): PSNR 4.2 dB, SSIM 0.007, VMAF 0.0 -- unusable.

### SPAN -- `2x_ModernSpanimationV2`
| res | dtype | fps before | fps after | Delta | PSNR | SSIM | VMAF |
|-----|-------|-----------:|----------:|------:|-----:|------|-----:|
| 360p  | fp16 | 82.6 | **104.8** | +27% | 62.9 | 0.9996 | -- |
| 360p  | fp32 | 39.2 | **44.7**  | +14% | 62.9 | 0.9996 | -- |
| 720p  | fp16 | 24.9 | **26.8**  | +7%  | 61.7 | 0.9995 | 98.2 |
| 720p  | fp32 | 11.1 | **11.6**  | +4%  | 61.7 | 0.9995 | -- |
| 1080p | fp16 | 11.7 | **12.1**  | +4%  | 61.5 | 0.9994 | -- |

PSNR/SSIM/VMAF are measured as fp16 output vs fp32 output of the same model (the
fp16 noise floor): unchanged before->after for PLKSR/SPAN, fixed for NAFNet -- no
quality regression beyond fp16 noise. SPAN's lower PSNR reflects its inherent
fp16 sensitivity, identical in both builds.

## Notes
- All three still export via `torch.onnx.export` (opset 20, dynamic axes).
- TensorRT FP16 before/after is neutral (PLKSR 720p 4.39->4.32 fps; SPAN exports
  byte-identical) -- PyTorch-eager wins with no downside on compiled paths.

### CODE REVIEW

**Comment by NevermindNilas** [comment]
Thanks for the review. Addressed in 67530e5:

- **SPAN dtype/device caching** — valid. `Conv3XC` now re-fuses `eval_conv` when `(dtype, device)` changes (flag initialized in `__init__`), so `.half()/.float()/.cuda()` after the first forward can't reuse a stale fusion. This matches upstream's per-forward behaviour on a dtype change while still skipping the redundant recompute within a dtype. (Note: a `.half()`→`.float()` round-trip on one instance is still limited to fp16 precision because the source conv weights were already rounded — that's inherent and identical on `main`.)
- **bench argparse `choices` + docstring filename** — fixed.
- **PLKSR ndim-generic affine** — done (`view(1, c, *1s)`).

- **PLKSR `channels_last` reshape** — this one is a false positive. `torch.reshape` operates on the *logical* (row-major NCHW) layout regardless of `memory_format`; for a `channels_last` tensor it inserts a contiguous copy, so groups are formed by channel, not by physical storage order. Verified numerically: `FastGroupNorm` vs `F.group_norm` on a `channels_last` fp16 tensor matches at PSNR ~68 dB, and the end-to-end fp16-vs-fp32 PSNR/SSIM in the benchmark table (79.5 dB / 0.99999) is computed on `channels_last` inputs. If it were grouping the wrong elements the output would be garbage.

**Comment by Sopel97** [comment]
I patched this in via 

```
def patch_spandrel_realplksr():
    import torch
    from torch import nn

    import spandrel.architectures.PLKSR.__arch.RealPLKSR as realplksr

    if getattr(realplksr, "_patched_fast_group_norm", False):
        return

    class FastGroupNorm(nn.GroupNorm):
        """GroupNorm with an fp32-accumulated inference fast path.

        PyTorch's native fp16 ``group_norm`` kernel is ~6-8x slower than a manual
        reduction for these shapes and forces a contiguous copy every call. In eval
        this computes the per-group mean/variance with fp32 accumulation (matching
        the native numerics) while keeping the elementwise normalize in the input
        dtype. Drop-in for ``nn.GroupNorm``: identical parameters and state_dict
        keys, and ONNX-exportable (ReduceMean / Sub / Rsqrt / Mul).

        Training defers to ``F.group_norm`` so the fused native backward (faster,
        less memory) is used -- the manual path is an inference-only optimization.
        """

        def forward(self, x: torch.Tensor) -> torch.Tensor:
            if self.training:
                return nn.functional.group_norm(
                    x, self.num_groups, self.weight, self.bias, self.eps
                )
            n, c = x.shape[0], x.shape[1]
            g = self.num_groups
            xr = x.reshape(n, g, -1)
            # Mean/centering/variance/normalize all in fp32 to match the native
            # numerics exactly, then cast back. The affine runs in the input dtype
            # on already-normalized (unit-scale) values, so no precision is lost.
            mean = xr.mean(2, keepdim=True, dtype=torch.float32)
            d = xr.float() - mean
            var = d.square().mean(2, keepdim=True)
            d = d * torch.rsqrt(var + self.eps)
            x = d.reshape(x.shape).to(x.dtype)
            if self.weight is not None:
                shape = (1, c) + (1,) * (x.ndim - 2)
                x = x * self.weight.view(shape) + self.bias.view(shape)
            return x

    # Replace PLKBlock's constructor with one that uses FastGroupNorm.
    original_init = realplksr.PLKBlock.__init__

    def patched_init(
        self,
        dim,
        kernel_size,
        split_ratio,
        use_ea=True,
        norm_groups=4,
        use_layer_norm=False,
    ):
        original_init(
            self,
            dim,
            kernel_size,
            split_ratio,
            use_ea,
            norm_groups,
            use_layer_norm,
        )

        if not use_layer_norm:
            old_norm = self.norm

            new_norm = FastGroupNorm(
                old_norm.num_groups,
                old_norm.num_channels,
                eps=old_norm.eps,
                affine=old_norm.affine,
            )

            if old_norm.affine:
                with torch.no_grad():
                    new_norm.weight.copy_(old_norm.weight)
                    new_norm.bias.copy_(old_norm.bias)

            self.norm = new_norm

    realplksr.PLKBlock.__init__ = patched_init
    realplksr._patched_fast_group_norm = True

patch_spandrel_realplksr()
```

and got around 5-10% speedup in my pipeline, thanks
